### PR TITLE
Removed buildapi and imported buildapi_client

### PR DESCRIPTION
**Pull request by F3real**

### CODE REVIEW

File: mozci/ci_manager.py [hunk 1]
```diff
@@ -10,12 +10,17 @@
 from __future__ import absolute_import
 
 from abc import ABCMeta, abstractmethod
+from buildapi_client import (
```
Comment (armenzg) [inline mozci/ci_manager.py:14]: Could you please add a white space above? We try to group 3rd party modules versus standard libraries like "abc". More comments to come.

File: mozci/mozci.py [hunk 2]
```diff
@@ -1,18 +1,13 @@
...
-you will be better off letting mozci.py determine which source to reach
-to take the actions you need.
-
 In here, you will also find high level functions that will do various low level
 interactions with distinct modules to meet your needs.
 """
 from __future__ import absolute_import
 
 import logging
+from buildapi_client import make_retrigger_request, trigger_arbitrary_job
```
Comment (armenzg) [inline mozci/mozci.py:11]: Could you please a white line above? Same reason about separating visually 3rd party modules.

File: mozci/mozci.py [hunk 3]
```diff
@@ -21,7 +16,7 @@
     is_downstream,
     list_builders,
 )
-from mozci.sources import buildapi, buildjson, pushlog
+from mozci.sources import repositories, buildjson, pushlog
```
Comment (armenzg) [inline mozci/mozci.py]: Could you please move repositories at the end of the list?
I try to keep them sorted alphabetically.

My apologies for this small nits! I know they can be annoying but it helps in the long term.

File: mozci/mozci.py [hunk 4]
```diff
@@ -52,7 +47,6 @@ def disable_validations():
     global VALIDATE
     if VALIDATE:
         LOG.debug("Disable validations.")
-        buildapi.VALIDATE = False
```
Comment (armenzg) [inline mozci/mozci.py:55]: Is there a way we could keep this feature?
Did we loose that in the buildapi_client release?

File: mozci/query_jobs.py [hunk 5]
```diff
@@ -3,10 +3,11 @@
 import logging
 
 from abc import ABCMeta, abstractmethod
+from buildapi_client import query_jobs_schedule
```
Comment (armenzg) [inline mozci/query_jobs.py:7]: Empty line above. Thanks!

File: mozci/scripts/trigger.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 import urllib
 
 from argparse import ArgumentParser
+from buildapi_client import make_retrigger_request
```
Comment (armenzg) [inline mozci/scripts/trigger.py:7]: Same thing. I will stop making this comment to not become annoying :+1:

File: mozci/sources/repositories.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+#! /usr/bin/env python
+"""
+This script is designed to trigger jobs through Release Engineering's
+buildapi self-serve service.
+
+The API documentation is in here (behind LDAP):
+https://secure.pub.build.mozilla.org/buildapi/self-serve
+
+The docs can be found in here:
+http://moz-releng-buildapi.readthedocs.org
```
Comment (armenzg) [inline mozci/sources/repositories.py]: Could you please adjust this paragraph?
We can talk about this being a way to determine meta-data about repositories.

File: mozci/mozci.py (hunk 4 above)
Comment (F3real) [inline mozci/mozci.py:55]: We removed pushlog.valid_revision  from buildapi when we moved it to buildapi_client.

File: mozci/sources/repositories.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+"""
+from __future__ import absolute_import
+import json
+import logging
+import os
+
+from buildapi_client import BuildapiAuthError, make_query_repositories_request
+
+from mozci.errors import BuildapiError, AuthenticationError
```
Comment (armenzg) [inline mozci/sources/repositories.py]: Could you remove BuildapiError from the errors module?
We should have that one in buildapi_client unless we forgot.

Comment (F3real) [inline mozci/sources/repositories.py]: I removed it because it wasn't used anywhere in buildapi_client.

File: mozci/sources/repositories.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+from mozci.utils.authentication import get_credentials, remove_credentials
+from mozci.utils.transfer import path_to_file
+
+HOST_ROOT = 'https://secure.pub.build.mozilla.org/buildapi/self-serve'
```
Comment (armenzg) [inline mozci/sources/repositories.py]: I don't think you're using this variable in here.

You can use this to catch more issues: "pylint -E mozci/sources/repositories.py"

File: mozci/sources/repositories.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
+#! /usr/bin/env python
```
Comment (armenzg) [inline mozci/sources/repositories.py]: Could you please this module outside of sources/?
Put it under mozci/
I try to keep sources/ for modules that interact with external services.

File: mozci/sources/repositories.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Return dictionary with information about a specific repository.
+
+    Raises BuildapiError if the repository does not exist.
+    """
+    repositories = query_repositories()
+    if repo_name not in repositories:
+        repositories = query_repositories(clobber=True)
+        if repo_name not in repositories:
+            raise BuildapiError("That repository does not exist.")
```
Comment (armenzg) [inline mozci/sources/repositories.py]: Raise a MozciError instead.

File: mozci/sources/repositories.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    if os.path.exists(REPOSITORIES_FILE):
+        LOG.debug("Loading %s" % REPOSITORIES_FILE)
+        fd = open(REPOSITORIES_FILE)
+        REPOSITORIES = json.load(fd)
+    else:
+        try:
+            REPOSITORIES = make_query_repositories_request(auth=get_credentials(), dry_run=False)
+        except BuildapiAuthError:
+            remove_credentials()
+            raise AuthenticationError("Your credentials were invalid. Please try again.")
```
Comment (armenzg) [inline mozci/sources/repositories.py]: Good. I like it.

File: test/test_query_jobs.py [hunk 13]
```diff
@@ -125,14 +125,13 @@ class TestBuildApiGetAllJobs(unittest.TestCase):
 
     def setUp(self):
         self.query_api = BuildApi()
-        buildapi.JOBS_CACHE = {}
+        repositories.JOBS_CACHE = {}
```
Comment (armenzg) [inline test/test_query_jobs.py]: From reading the code, I don't think we have have a JOBS_CACHE for repositories.

File: test/test_query_jobs.py [hunk 14]
```diff
@@ -125,14 +125,13 @@ class TestBuildApiGetAllJobs(unittest.TestCase):
...
         query_jobs.JOBS_CACHE = {}
 
     @patch('requests.get', return_value=mock_response(JOBS_SCHEDULE, 200))
-    @patch('mozci.sources.pushlog.valid_revision', return_value=True)
-    @patch('mozci.sources.buildapi.get_credentials', return_value=None)
-    @patch('mozci.sources.buildapi.query_repo_url', return_value=None)
-    def test_call_first_time(self, query_repo_url, get_credentials, valid_revision, get):
+    @patch('mozci.query_jobs.get_credentials', return_value=None)
```
Comment (armenzg) [inline test/test_query_jobs.py:130]: Does patching this module instead works? If not, no worries.
https://github.com/mozilla/mozilla_ci_tools/blob/master/mozci/utils/authentication.py#L74

File: tox.ini [hunk 15]
```diff
@@ -2,7 +2,7 @@
 envlist = py27
 
 [testenv]
-basepython = python2.7
+basepython = python
```
Comment (armenzg) [inline tox.ini:5]: Why did you need to change this to "python"? Was it giving you trouble?

File: mozci/sources/repositories.py (hunk 8 above)
Comment (armenzg) [inline mozci/sources/repositories.py]: I made a comment later on that we should use MozciError instead.

File: tox.ini (hunk 15 above)
Comment (F3real) [inline tox.ini:5]: Tox fails to find python for me otherwise, but I am on Windows so that maybe issue.

**Comment by armenzg** [comment]
Hi @F3real this is going in the right direction. Good job!

Once you're done making all fixes, please make a comment in here and if I have time today I will look at it again. If not on Monday.

Before merging this PR, I would like to squash all the commits into 1. That way when we merge it we have only one big diff which is easier to read and understand everything that is changing. Rather than loading various commits.

Could you please read this and let me know if it makes sense?
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html
Use "git diff master... > ~/back_up_of_changes.diff" to keep a copy of your changes when trying to squash your commits. Just in case you get into trouble.
If you're not confortable doing it, I could give you a hand learning how to do it.

File: test/test_query_jobs.py (hunk 14 above)
Comment (F3real) [inline test/test_query_jobs.py:130]: It doesn't for some reason.

**Comment by armenzg** [comment]
Woohoo!
